File: content/排序_希尔排序/shellSort.cpp
/*
时间复杂度与最优步长序列有关，最优为O(nlog^2 n)，最差为O(n^2)
空间复杂度为 O(n)
*/

#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

void shellSort(vector<int> &array)
{
  int h = 1;
  while (h < array.size() / 3)
  {
    h = h * 3 + 1;
  }
  while (h >= 1)
  {
    for (int i = h; i < array.size(); i++)
    {
      for (int j = i; j >= h && array[j - h] > array[j]; j -= h)
      {
        swap(array[j - h], array[j]);
      }
    }
    h /= 3;
  }
}

int main()
{
  vector<int> a = {1, 3, 2, 4, 6, 7, 5, 9, 10, 8};
  shellSort(a);
  for (auto i : a)
  {
    cout << i << endl;
  }
  return 0;
}

File: content/shellSort/ShellSort.java
/*
 * @Author: your beardada
 * @Date: 2021-04-09 08:27:31
 * @LastEditTime: 2021-04-12 08:26:16
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit
 * @FilePath: \content\shellSort\ShellSort.java
 */
package shellSort;

public class ShellSort {
  public static void shellSort(int[] list) {
    int n = list.length / 3;
    int h = 1;
    while (h < n) {
      h = h * 3 + 1;
    }
    while (h >= 1) {
      for (int i = h; i < list.length; i++) {
        for (int j = i; j >= h && list[j] < list[j - h]; j -= h) {
          swap(list, j, j - h);
        }
      }
      h /= 3;
    }
  }

  public static void swap(int[] list, int i, int j) {
    int temp = list[i];
    list[i] = list[j];
    list[j] = temp;
  }

  public static void print(int[] list) {
    for (int i : list) {
      System.out.println(i);
    }
  }

  public static void main(String[] args) {
    int[] a = { 1, 4, 5, 6, 3, 2, 0, 9, 8, 7 };
    shellSort(a);
    print(a);
  }
}
